doc/source/data/key-concepts.rst

```diff
@@ -156,3 +258,53 @@ You can tell if stage fusion is enabled by checking the :ref:`Dataset stats <dat
     * Remote wall time: T min, T max, T mean, T total
     * Remote cpu time: T min, T max, T mean, T total
     * Output num rows: N min, N max, N mean, N total
+
+Block Move Semantics
+~~~~~~~~~~~~~~~~~~~~
+
+In addition to fusing together stages, lazy execution mode further optimizes memory
+utilization by eagerly releasing the data produced by intermediate operations in a
+chain.
+
+For example, if you have a chain of ``read() -> map() -> filter()`` operations:
+
+.. code-block::
+
+    ds = ds.read_parquet().experimental_lazy().map(udf).filter(filter_udf)
```

Comment:
This is obvious pseudo-code, so I think this is ok. But then again, can we have a self-contained snipped that runs? If so, can we test and include it?